lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+//
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdTransform3dMayaXformStack.h"
+
+#include <mayaUsd/ufe/Utils.h>
+
+#include <mayaUsd/fileio/utils/xformStack.h>
+
+#include "private/UfeNotifGuard.h"
+
+#include <maya/MEulerRotation.h>
+
+#include <map>
+
+namespace {
+
+using namespace MayaUsd::ufe;
+
+// Type traits for GfVec precision.
+template<class V>
+struct OpPrecision {
+    static UsdGeomXformOp::Precision precision;
+};
+
+template<>
+UsdGeomXformOp::Precision OpPrecision<GfVec3f>::precision = 
+    UsdGeomXformOp::PrecisionFloat;
+
+template<>
+UsdGeomXformOp::Precision OpPrecision<GfVec3d>::precision = 
+    UsdGeomXformOp::PrecisionDouble;
+
+inline double TO_DEG(double a) { return a * 180.0 / 3.141592654; }
+inline double TO_RAD(double a) { return a * 3.141592654 / 180.0; }
+
+VtValue getValue(const UsdAttribute& attr, const UsdTimeCode& time)
+{
+    VtValue value;
+    attr.Get(&value, time);
+    return value;
+}
+
+// UsdMayaXformStack::FindOpIndex() requires an inconvenient isInvertedTwin
+// argument, various rotate transform op equivalences in a separate
+// UsdMayaXformStack::IsCompatibleType().  Just roll our own op name to
+// Maya transform stack index position.
+const std::unordered_map<std::string, UsdTransform3dMayaXformStack::OpNdx> opNameToNdx{
```

Comment:
Op name to Maya transform stack index position mapping.  This is what we use to sort, and to determine all transform ops that can be supported as rotation.